pkg/routeagent_driver/handlers/calico/ippool_handler.go

```diff
@@ -193,3 +207,79 @@ func (h *calicoIPPoolHandler) deleteIPPool(endpoint *submV1.Endpoint) error {
 func getEndpointSubnetIPPoolName(endpoint *submV1.Endpoint, subnet string) string {
 	return fmt.Sprintf("submariner-%s-%s", endpoint.Spec.ClusterID, strings.ReplaceAll(subnet, "/", "-"))
 }
+
+func (h *calicoIPPoolHandler) platformIsROKS() bool {
+	// Submariner GW is deployed on ROKS using LB service with specific annotations.
+	service, err := h.k8sClient.CoreV1().Services(h.namespace).Get(context.TODO(), gwLBSvcName, metav1.GetOptions{})
+	if err != nil {
+		return false
+	}
+
+	if service.GetAnnotations()[gwLBSvcROKSAnnotation] != "" {
+		return true
+	}
+
+	return false
+}
+
+// workaround to address datapath issue with default Calico IPPool configuration for ROKS platform,
+// IPIPMode of calico default IPPool should be set to 'Always'.
+func (h *calicoIPPoolHandler) updateROKSCalicoCfg() error {
+	if !h.platformIsROKS() {
+		return nil
+	}
+
+	// platform is ROKS, make sure that IPIPMode of default IPPool is Always
+	defaultIPPool, err := h.client.ProjectcalicoV3().IPPools().Get(context.TODO(), defaultV4IPPoolName, metav1.GetOptions{})
+	if err != nil {
+		return errors.Wrapf(err, "error reading %s IPPool", defaultV4IPPoolName)
+	}
+
+	if defaultIPPool.Spec.IPIPMode != calicoapi.IPIPModeAlways {
+		if defaultIPPool.Labels == nil {
+			defaultIPPool.Labels = make(map[string]string)
+		}
+
+		if defaultIPPool.Annotations == nil {
+			defaultIPPool.Annotations = make(map[string]string)
+		}
+		// mark 'user-managed' label for persistency
+		defaultIPPool.Labels["user-managad"] = "true"
+		defaultIPPool.Annotations["submariner-prev-ipipmode"] = string(defaultIPPool.Spec.IPIPMode)
+		defaultIPPool.Spec.IPIPMode = calicoapi.IPIPModeAlways
+		_, err = h.client.ProjectcalicoV3().IPPools().Update(context.TODO(), defaultIPPool, metav1.UpdateOptions{})
+
```

Comment:
We should use admiral's `util.Update` which handles potential conflicts. That also does a `Get` and handles `IsNotFound` so we can remove the `Get` above. 

```
err = util.Update(context.TODO(), iPPoolResourceInterface(), defaultIPPool, func(existing *calicoapi.IPPool) (*calicoapi.IPPool, error) {
    if existing.Spec.IPIPMode == calicoapi.IPIPModeAlways {
        return existing, nil // no need to update
    }

    if existing.Labels == nil {
	existing.Labels = make(map[string]string)
    }

    if existing.Annotations == nil {
        existing.Annotations = make(map[string]string)
    }

    // mark 'user-managed' label for persistency
    existing.Labels["user-managad"] = "true"
    existing.Annotations["submariner-prev-ipipmode"] = 
                             string(defaultIPPool.Spec.IPIPMode)
    existing.Spec.IPIPMode = calicoapi.IPIPModeAlways

    return existing, nil
})

...

func (h *calicoIPPoolHandler) iPPoolResourceInterface() resource.Interface[*calicoapi.IPPool] {
	return &resource.InterfaceFuncs[*calicoapi.IPPool]{
		GetFunc:    h.client.ProjectcalicoV3().IPPools().Get,
		CreateFunc: h.client.ProjectcalicoV3().IPPools().Create,
		UpdateFunc: h.client.ProjectcalicoV3().IPPools().Update,
		DeleteFunc: h.client.ProjectcalicoV3().IPPools().Delete,
	}
}
```

Reply:
I want to log the change in case we update the default IPPool configuration, do you know where I can do that with the Admiral's util?

Reply:
You can do that in your callback function (`func(existing *calicoapi.IPPool)`).